### PR TITLE
[FIX] stock: uninstalling stock app removes stock moves received qty to 0

### PR DESCRIPTION
**Pull request by peso-odoo**
After uninstalling the stock module on a database which also has the purchase (and `purchase_stock`) module installed, the `qty_received_method` is removed for purchase order lines handling the reception of the products through the stock module (`qty_received_method = 'stock_moves'`). Because of this, a recompute is triggered on the `qty_received`, setting it to zero.

This leaves the purchase order in an invalid state (the received quantity did not change through the uninstallation of the stock module), additionally the problem cannot be corrected, since the receiving method is not set to `manual`.

Functionally, the uninstallation shouldn't update the purchase orders, instead they should be decoupled from the associated stock moves. To this end, an `ondelete` handler is added for the `stock_moves` selection option.

Steps to reproduce:

 - Install Purchases app
 - Install Inventory app
 - Create a purchase order with purchase lines and quantity > 0
 - Confirm the purchase order
 - Click on receive products
 - Click on validate
 - Uninstall Inventory app
 - Check that the purchase order lines have the received field set to zero and it is not editable

### CODE REVIEW

**Comment by tdecaluwe** [comment]
@peso-odoo It might make more sense to wrap this logic in an `ondelete` handler instead of an uninstall hook:
```python
class PurchaseOrderLine(models.Model):
    def _ondelete_stock_moves(self):
        # Uninstall hook logic goes here

    qty_received_method = fields.Selection(selection_add=[('stock_moves', 'Stock Moves')], ondelete={
        'stock_moves': _ondelete_stock_moves
    })
```

As documented in https://www.odoo.com/documentation/15.0/developer/reference/backend/orm.html#odoo.fields.Selection

File: addons/stock/__init__.py [hunk 1]
```diff
@@ -29,3 +29,16 @@ def _assign_default_mail_template_picking_id(cr, registry):
         company_ids_without_default_mail_template_id.write({
             'stock_mail_confirmation_template_id': default_mail_template_id.id,
         })
+
+def _uninstall_hook(cr, registry):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    try:
```
Comment (MiquelRForgeFlow) [inline addons/stock/__init__.py:35]:
```suggestion
    if not "purchase.order.line" in env.registry.models:
        return
    try:
```
avoid KeyError in case purchase is not installed 🤔

File: addons/stock/__init__.py [hunk 2]
```diff
@@ -19,6 +18,7 @@ def pre_init_hook(cr):
         ('module', '=', 'stock')
     ]).unlink()
 
+
```
Comment (nle-odoo) [inline addons/stock/__init__.py:21]: adding or removing newlines in a file without change is an unnecessary diff

to remove it you can checkout the file to previous version (`git checkout HEAD~1 -- addons/stock/__init__.py`) and amend then push

File: addons/purchase_stock/models/purchase.py [hunk 3]
```diff
@@ -272,7 +272,13 @@ def _get_orders_to_remind(self):
 class PurchaseOrderLine(models.Model):
     _inherit = 'purchase.order.line'
 
-    qty_received_method = fields.Selection(selection_add=[('stock_moves', 'Stock Moves')])
+    def _ondelete_stock_moves(self):
+        for record in self:
+            record.qty_received_manual = record.qty_received
+            record.qty_received_method = 'manual'
```
Comment (rco-odoo) [inline addons/purchase_stock/models/purchase.py:278]: Using the `ondelete` mechanism is quite nice (it was designed for that purpose, after all.)
However, I am wondering about performance: on a large database, this may never end.
What about doing it in plain SQL instead?
The only potential issue would be dependent computed fields that would not be recomputed...

Comment (tdecaluwe) [reply addons/purchase_stock/models/purchase.py:278]: Yes, I thought about this, but you're also mentioning the exact reason why I didn't. I thought about it some more though, and I'm wondering about these options:
1) What if we updated `qty_received_manual` through sql? And then do a single `write` to set the `qty_received_method`? Because the only compute depending on `qty_received_method` also depends on `qty_received_manual` (it is the field `qty_received`) we could probably use the following strategy to avoid a linear amount of queries: simply check if the computed value is already correct, if it is don't write it.
    ```python
    @api.depends('qty_received_method', 'qty_received_manual')
    def _compute_qty_received(self):
        lines_manual = self.filtered(lambda line: (
            line.qty_received_method == 'manual'
            and line.qty_received != line.qty_received_manual
        ))
        # This is a zero-length for loop when triggering the sql based ondelete.
        for line in lines_manual:
            line.qty_received = line.qty_received_manual or 0.0

        self.filtered(lambda line: line.qty_received_method != 'manual').write({
            'qty_received': 0.0,
        })
    ```
2) Can we trigger computes manually? If yes, combining this with the strategy above ensures correctness in all cases (including non-odoo custo computes depending on these fields).

File: addons/purchase/models/purchase.py [hunk 4]
```diff
@@ -1068,11 +1068,16 @@ def _compute_qty_received_method(self):
...
+        lines_manual = self.filtered(lambda line: (
+            line.qty_received_method == 'manual'
+            and line.qty_received != line.qty_received_manual
+        ))
+        for line in lines_manual:
+            line.qty_received = line.qty_received_manual or 0.0
+
+        self.filtered(lambda line: line.qty_received_method != 'manual').write({
+            'qty_received': 0.0,
+        })
```
Comment (rco-odoo) [inline addons/purchase/models/purchase.py:1080]: OMG please don't use `write` to assign a field in a compute method.
```suggestion
        self.filtered(lambda line: line.qty_received_method != 'manual').qty_received = 0.0
```

[the remark this answers is not in the document]
Comment (tdecaluwe) [reply addons/purchase/models/purchase.py]: The goal was to remove the loop over the non-manual lines. I updated the code to make it simpler. It seems the extra if is not even necessary. Does the cache perhaps perform the same check (no writes if `new_value == old_value`)?

**Comment by amoyaux** [comment]
Seems okay for stable but IMO it should not land in master. qty_received_manual seems to be useless since we can now do compute field with a mix of automatic and manual update. On top of it you also have the issue with the installation of the module. I think this code need a bit of cleaning to avoid this kind of issues (either you do it or we can also do it if you don't have time for that)

Also just for curiosity, are there any differences between ondelete and uninstall_hook? Because I prefer to add the complexity of uninstallation outside the file and try to keep it the more simple possible (personal preference).

**Comment by amoyaux** [comment]
robodoo r+

**Comment by amoyaux** [comment]
Did you rebase or fix? Because the runbot fail is a consequence of this PR and not a false positive.

For the issue:
- Create a PO for a product consumable
- Confirm the PO
- Install stock

There is no move for the PO line and you can't edit the quantities

**Comment by nle-odoo** [comment]
> Did you rebase or fix? Because the runbot fail is a consequence of this PR and not a false positive

From the traceback, it's possible that the added test works in community, but fail with enterprise (there is lots of "quality_control" in the error that is in enterprise). But in any case the test should work even if enterprise is installed.

**Comment by tdecaluwe** [comment]
The purchase user (`self.res_users_purchase_user`) has writing rights on `stock.move` similarly to stock users, however the purchase user group does not imply the rights on `quality.point` as the stock user group does. Not exactly sure what the best solution is. I'm guessing we want the purchase user to be able to write on `quantity_done` with minimum quality control rights?

**Comment by nle-odoo** [comment]
robodoo r+

(fix of errored test seemed to be to admin user run the test, which makes sense if this is for uninstallation)
